content/blog/nielx/2022-12-16_http_network_services_preview_in_r1_beta_4.md

```diff
@@ -0,0 +1,109 @@
+
+The Preview of the HTTP Network Services Kit uses some of the C++ language and library features that are currently (mostly) unused in the rest of the project. Most of Haiku follows the design standards originating from Be, which means that most of the API and the code is based on the (pre-) C++98 version of the language. Furthermore, because of Haiku's promise of binary compatiblity, most of the code will still have to build on GCC 2.95.3, which was released on 16 March 2001. This compatibility guarantee is not there for any new APIs and kits. While the new Network Services Kit is in development, it will include some experiments in its design and use of the language.
+
+This section highlights three of those changes: the use of modern C++ language and library features, the introduction of exceptions for error handling, and an experiment in data access and control mechanisms through exclusive borrows.
+
+### Modern C++ (C++17)
+
+| Feature | Example | Description |
+| ------- | ------- | ----------- |
+| **Move Semantics** | [`BHttpSession::Execute()`](https://git.haiku-os.org/haiku/tree/src/kits/network/libnetservices2/HttpSession.cpp?h=r1beta4#n219) | Arguably [move semantics](https://www.cprogramming.com/c++11/rvalue-references-and-move-semantics-in-c++11.html) is the largest new feature in the language in C++11. The libnetservices2 kit utilizes move semantics to the fullest, meaning in the implementation all classes abide by the [rule of six](https://www.modernescpp.com/index.php/c-core-guidelines-constructors-assignments-and-desctructors), which makes it possible to pass objects by value in most cases. The example of the [`Execute()`](https://www.haiku-os.org/docs/api/classBPrivate_1_1Network_1_1BHttpSession.html#ae14a7c52aaddfd69e22be4a8aeff727f) method, is the only place where the new API forces you to pass the request object as an r-value reference. |
+| **Range-based Iteration** | [`BHttpFields`](https://git.haiku-os.org/haiku/tree/src/kits/network/libnetservices2/HttpFields.cpp?h=r1beta4#n477) | The [range-based `for` loop](https://learn.microsoft.com/en-us/cpp/cpp/range-based-for-statement-cpp?view=msvc-170) syntax is supported by the `BHttpFields` class. This allows you to loop through all the header fields in the object using this construct. Internally, it is used in most cases where there are loops through containers. |
+| **String Views** | [`BHttpMethod::BHttpMethod()`](https://git.haiku-os.org/haiku/tree/src/kits/network/libnetservices2/HttpRequest.cpp?h=r1beta4#n70) | The [`std::string_view`](https://en.cppreference.com/w/cpp/string/basic_string_view) object is a very simple concept: it is a pointer to a string, with a length embedded. This makes the use of borrowing (parts of) strings a lot simpler and safer, and has a lot of potential in the rest of the API. It is probably a good opportunity to investigate if this can be supported in the Haiku API more (or if it is needed to create a custom variant). |
```

Comment:
BString already has copy-on-write but I don't know if it is agood as what string view does. In any case, we could probably change the internal design of BString, instead of ending up with two classes (string_view and string) in the standard C++, favorizing ease of use of the API over performance

Reply:
A `string_view` is a ... view on a string. I.e. it points to the start of the data buffer, and has a length. This mostly makes it a safe(r) and more efficient variant of const char* way of passing pointers to strings around. Additionally, it implements some methods to do some comparisons with std::string objects. I think for reasons of safety and versatility, we should investigate if we want to add the ability to pass a string view like object to methods where now we accept const char*. We should see if we want to reuse std::string_view, or whether there will be a BStringView.

Your point that a BString can be used as an alternative because of the COW design is valid and might be the answer, though one downside is that it still relies on the internal object to be nul-terminated, whereas a string_view by virtue of the fact that it packs a pointer + length does not have that requirement and therefore can also point to substrings within larger data buffers.

Interesting discussion, though for the mailing list/comments on the article. Anything you think I could change in the text to make it clearer?

Reply:
I am familiar with string_view, thanks.

The problem with std::string is that it has a .c_str() method that returns a const char* pointer to its internal buffer. This means essentially free conversion from std::string to const char*. But it also means the buffer has to be null-terminated at exactly the right place.

You could design a complete API without a thing like std::string and only ever have string_views. If we can, I would prefer that BString goes that way. It means more expensive conversion from BString to const char*, so care must be taken to not need const char* so much. And, like the c++ standard strings, we may be limited by ABI/API stability and decide to preserve both null-terminated and lenght-based strings.

Anyways, future APIs aside, I guess it makes sense to have string_view in an heavily text oriented protocol like http, and it is certainly a better idea than strtok we were using before. So, maybe we can just say that high-performance code here requires specific tools, and BString just won't cut it?